docs: 📝 Add article about color generation (#2976)
digdir/designsystemet: apps/storefront/app/bloggen/2025/generering-av-fargar-i-fargesystemet/page.mdx

* Hue (fargetone) beskriver sjølve fargen og angis på en skala frå 0° til 360°, som for eksempel raud, grøn eller blå. 
* Saturation (metning) angir kor intens eller grålig fargen er frå ein skala frå 0 %  til 100 %. 
* Lightness (lysstyrke) bestemmer kor lys eller mørk fargen er, frå 0 % (svart) til 100 % (hvit).

For å lage fargeskalaer i HSL-fargerommet finst det fleire teknikkar som vert nytta.
I eksempla nedover vil me sjå på nokre av desse teknikkane og korleis dei fungerer.

 <Image
  src='/img/bloggen/fargar-hsl.png'
  alt='Illustrasjon av HSLuv1.'
  caption='Viser ein fargeskala med utgangspunkt i Digdir sin raude profilfarge i HSL-fargerommet.'
  boxShadow={false}
  dataUnstyled
/>

Ein vanleg metode er å justere ligthtness (lysstyrka) for ein farge for å gjere den mørkare eller lysare. 
I bildet over har me tatt utgangspunkt i den raude profilfargen til Digdir (`67% lightness`) og satt faste lightness-verdiar for å lage ulike valører.
Samenlikna med RGB-teknikken, så ser me at fargane blei mykje meir saturerte oppover i skalaen. 
Dette gjer mening fordi me har ikkje justerte på satueringsverdiane. 

> Dette illustrerar veldig godt ein utfordring med mange av fargeromma som brukar lightness for å definere lysstyrka i fargen.

Reviewer: ```suggestion
Dette illustrerar veldig godt ei utfordring med mange av fargeromma som brukar lightness for å definere lysstyrka i fargen.
```

> Endrar me kun på denne verdien så endar fargane veldig ofte opp med å sjå oversaturerte og ubalanserte ut.

Reviewer: ```suggestion
Endrar me kun på denne verdien så endar fargane ofte opp med å sjå oversaturerte og ubalanserte ut.
```
Unødvendig ord, treng ikkje bruke ordet "veldig" for mykje

 <Image
  src='/img/bloggen/fargar-hsl-2.png'
  alt='Illustrasjon av HSLuv1.'
  caption='Viser ein fargeskala med utgangspunkt i Digdir sin raude profilfarge i HSL-fargerommet med redusert saturation (fargemetning).'
  boxShadow={false}
  dataUnstyled
/>

Ein anna metode for å lage fargeskalaer i HSL-fargerommet er å justere på både saturation og lightness verdiane.

> I bildet over har me fortsatt tatt utgangspunkt i den raude profilfargen til Digdir med faste lightness-verdiar.

Reviewer: ```suggestion
I biletet over har me fortsatt tatt utgangspunkt i den raude profilfargen til Digdir med faste lightness-verdiar.
```

> I tillegg har me satt saturation verdien ned frå `87` til `65` på alle fargene (profilfargen til Digdir hadde `87` saturation). 

Reviewer: ```suggestion
I tillegg har me satt saturation verdien ned frå `87` til `65` på alle fargane (profilfargen til Digdir hadde `87` saturation). 
```

Å redusere saturation-verdien på alle fargane hjelper, men problemet er at me då mister den opphavlege fargen som blei brukt til å generere fargeskalaen. 
For oss var det viktig at denne fargen fortsatt var med, sidan den ofte kan vere ein viktig farge for brukaren.

Faste saturation-verdiar på tvers av fargane fungerte heller ikkje så godt i praksis. 
Me fann ut at det måtte skrivast veldig mykje komplisert logikk for å få fargar med ulike fargetonar til å fungere godt i saman.

> Her var RGB-teknikken mykje enklare å jobbe med, og laga mykje bedre fargar.

Reviewer: ```suggestion
Her var RGB-teknikken mykje enklare å jobbe med, og laga mykje betre fargar.
```

## Kva med dei andre fargeromma?
Etter å ha testa ut RGB og HSL fargeromma, så var det på tide å sjå på kva dei andre fargeromma hadde å by på. 

 <Image
  src='/img/bloggen/fargerom-digdir-rod.png'
  alt='Illustrasjon av HSLuv1.'
  caption='Eksempel på fargeskalaer i ulike fargerom.'
  boxShadow={false}
  dataUnstyled
/>

I bildet over ser du fargeskalaer med utgangspunkt i Digdir sin raude profilfarge i ulike fargerom. 
Mange av desse fargeromma brukar lightness (lysstyrke) for å definere kor lys ein farge er. 
Me opplevde derfor mange av dei samme utfordringane med desse fargeromma som med HSL, 
der me måtte skrive ein del kode for å få fargar med ulike fargetonar til å fungere godt i saman.

> Eit fargerom som me var veldig fornøydde med var OKLCH-fargerommet. 

Reviewer: ```suggestion
Eit fargerom som me var veldig fornøgde med, var OKLCH-fargerommet. 
```

Det laga ganske fine og balanserte fargar, men me opplevde at fargane av og til blei litt for saturerte for enkelte fargetonar. 

> Konklusjonen etter å ha vurdert alle desse fargeromma var at RGB-fargerommet viste seg å vere den enklaste og beste metoden for å lage 

Reviewer: ```suggestion
Konklusjonen etter å ha vurdert alle desse fargeromma, var at RGB-fargerommet viste seg å vere den enklaste og beste metoden for å lage 
```

> fargeskalaer på som ga oss dei fargane me var ute etter.

Reviewer: ```suggestion
fargeskalaer på, som ga oss dei fargane me var ute etter.
```

Men korleis kunne me sy dette saman i eit system som genererer fargeskalaer basert på kontrast og tilgjengeligheit?

## WCAG og offentlege nettsider 

> Vår tilnærming til utvkinga av fargesystemet er å sikre at alt me gjer oppfyller alle AA-krava i WCAG.

Reviewer: ```suggestion
Vår tilnærming til utviklinga av fargesystemet er å sikre at alt me gjer oppfyller alle AA-krava i WCAG.
```
Burde me legge til lenkje til AA-krava også?

Samstundes ønskte me å strekke oss mot AAA-nivået der det var meiningsfullt og ikkje gjekk utover designet.

> For å oppnå AA-krava må det vere `4.5:1` kontrast mellom tekst og bakgrunn for små tekstar og `3:1` for tekstar som er større eller lik `18 px`.

Reviewer: ```suggestion
For å oppnå AA-krava må det vere `4.5:1` kontrast mellom tekst og bakgrunn for små tekstar, og `3:1` for tekstar som er større eller lik `18 px`.
```

For interaktive og grafiske element er grensa `3:1`. Per 2025 seier regelverket at offentlege nettsider må oppfylle alle AA-krava i WCAG. 

> ### Korleis WCAG kalkulerer kontrast

Reviewer: ```suggestion
### Korleis WCAG kalkulerar kontrast
```

Når WCAG 2 kalkulerer kontrasten mellom to fargar, så bruker dei ein formel som tek utgangspunkt i den relative luminansen 
mellom fargane og gir dei ein poengsum. 
Denne poengsummen indikerer kor god kontrasten er. Har fargane best mogleg kontrast, er poengsummen `21:1`,

> medan om fargane ikkje har kontrast i det heile tatt, er poengsummen `1:1`. 

Reviewer: ```suggestion
om fargane ikkje har kontrast i det heile tatt, er poengsummen `1:1`. 
```

Luminansen skildrar den opplevde lysstyrka til ei flate og blir brukt til å vurdere kor mykje lys som blir reflektert

> eller sendt ut i ein bestemt retning. 

Reviewer: ```suggestion
eller sendt ut i ei bestemt retning. 
```

> Relativ luminans er ein normalisert verdi som rangerer lysstyrka til ein farge mellom rein svart `0` og rein kvit `1`. 

Reviewer: ```suggestion
Relativ luminans er ein normalisert verdi som rangerer lysstyrken til ein farge mellom rein svart `0` og rein kvit `1`. 
```

> Denne verdien blir rekna ut basert på dei raude, grøne og blå komponentane i fargen, der grønt bidreg mest til den opplevde lysstyrka. 

Reviewer: ```suggestion
Denne verdien blir rekna ut basert på dei raude, grøne og blå komponentane i fargen, der grøn bidreg mest til den opplevde lysstyrka. 
```

 <Image
  src='/img/bloggen/fargar-contrast.png'
  alt='Illustrasjon av HSLuv1.'
  caption='Illustrasjon som viser at kontrast og relativ luminans er like for 2 fargar.'
  boxShadow={false}
  dataUnstyled
/>

I bildet over er det 2 farga bakgrunnar med svart tekst i seg. 
Kontrasten mellom begge fargane og den svarte teksten er `6.6:1`, og den relative luminansen til dei farga bakgrunnen er `0.28`.
Det at desse 2 verdiane er like viser at WCAG brukar den relative luminansen mellom fargane for å kalkulere kontrasten.

## Fargerom basert på kontrast
Som nemnt tidlegare finst det mange fargerom som har ein lightness-verdi som viser kor lys ein farge er. 
I nokre av desse fargeromma betyr det at fargar med same lightness-verdi får om lag lik kontrast. 
To fargerom som oppnår dette er HSLuv og OKLCH.

### Fargeromma HSLUV og OKLCH
HSLuv og OKLCH lagar fargar som passar betre for korleis menneske ser dei. 

> HSLuv justerer fargane så dei held seg balanserte og inne i RGB-området, medan OKLCH brukar ein nyare modell som gir meir presise og jamne fargar. 

Reviewer: ```suggestion
HSLuv justerer fargane så dei held seg balanserte, også i RGB-området, medan OKLCH brukar ein nyare modell som gir meir presise og jamne fargar. 
```

Begge gjer det enklare å lage fine og harmoniske fargar, men OKLCH er meir nøyaktig.

 <Image
  src='/img/bloggen/hsluv-farger.png'
  alt='Illustrasjon av HSLuv1.'
  caption='Viser to fargeskalaer i OKLCH fargerommet.'
  boxShadow={false}
  dataUnstyled
/>

I bildet over vises to fargeskalaer i OKLCH-fargerommet som begge baserer seg på profilfargane til Digdir. 
Det som er veldig interessant et at om ein tar fargane med 70% lightness i skalaene over, så har fargane nesten identisk kontrast mot ein svart tekstfarge.

> Dette betyr i praksis at ein kan generere opp så mange fargeskalaer som ein vil, og sikre at kontrastane mellom dei ulike stega i fargeskalaene vil alltid vere like.

Reviewer: ```suggestion
Dette betyr i praksis at ein kan generere opp så mange fargeskalaer som ein vil, og sikre at kontrastane mellom dei ulike stega i fargeskalaene alltid vil være like.
```

OKLCH-fargerommet laga fine fargar og fungerte veldig godt for å sikre at kontrasten mellom fargar blei like på tvers av fargeskalaene.
Det som var litt synd var at fargane av og til blei litt for saturerte for enkelte fargetonar. 

Kva om me kunne kombinara RGB-fargerommet med OKLCH sin måte å sikre at kontrastane blei like på tvers av fargeskalaene? 
For å gjere dette så må me første snakke litt om interpolering.

## Interpolering av fargar i RGB-fargerommet
Interpolering er ein matematisk metode for å estimere verdiar mellom to kjente datapunkt. 

> Det blir ofte brukt når ein har eit datasett og ønskjer å finne verdiar som ligg mellom punktane. 

Reviewer: ```suggestion
Det blir ofte brukt når ein har eit datasett og ønskjer å finne verdiar som ligg mellom punkta. 
```

Interpolering av fargar handlar om å trekkje ei linje mellom to punkt i eit gitt fargerom. 
Resultatet varierer mykje avhengig av kva fargerom som blir brukt, og om linja er rett eller kurva. 

> For vår del handla det om å trekkje ei linje i gjennom RGB-fargerommet.

Reviewer: ```suggestion
For vår del handla det om å trekkje ei linje gjennom RGB-fargerommet.
```

 
 <Image
  src='/img/bloggen/interpolering.png'
  alt='Tre fargesirkler viser hvordan gradienter dannes ved ulike linjer trukket gjennom sirkelen: en rett linje, en buet linje nedover, og en buet linje oppover. Under hver sirkel vises tilhørende gradient som endrer seg basert på retningen og lengden på linjene. '
  caption='Biletet viser ei forenkla visualisering av fargeinterpolering der berre fargetone (hue) er med. Her ser ein korleis gradientane endrar seg avhengig av kvar linja vert trekt i sirkelen. '
  boxShadow={false}
  dataUnstyled
/>

Me visste allereie at WCAG brukte relativ luminans for å kalkulere kontrasten mellom to fargar. 

> For å sikre at kontrasten mellom to fargar alltid blir like, så må derfor den relative luminanse og vere lik.

Reviewer: ```suggestion
For å sikre at kontrasten mellom to fargar alltid blir like, så må den relative luminansen også være lik.
```

Det me ønskte å gjere var å finne luminansverdiar i RGB-fargerommet ved å "tinte" og "shade" fargane heilt til me oppnådde den luminansverdien me var ute etter.

> Å bevege seg iterativt gjennom eit fargerom på denne måten er ein form for interpolering.

Reviewer: ```suggestion
Å bevege seg iterativt gjennom eit fargerom på denne måten er ei form for interpolering.
```

 <Image
  src='/img/bloggen/rgb-interpolation.png'
  alt='Illustrasjon av HSLuv1.'
  caption='Viser eit eksempel på interpolering av fargar i RGB-fargerommet.'
  boxShadow={false}
  dataUnstyled
/>

I bildet over er det 2 fargar heilt til venstre med luminansverdiane `0.12` og `0.33`. 
Desse fargane fungerer som utgangspunkt for interpoleringen. 
Målet er å finne luminansverdien `0.7` for begge fargane i RGB-fargerommet, som du ser heilt til høgre.

> For å gjere dette ser me først på hva luminansverdiar me har å jobbe med i fargane heilt til venstre.

Reviewer: ```suggestion
For å gjere dette ser me først på kva luminansverdiar me har å jobbe med i fargane heilt til venstre.
```

Begge desse verdiane er lågare enn `0.7`. Me må derfor 'tinte' fargane for å få dei til å nærme seg `0.7`.

I eksemplet over tintar me fargane med 10% om og om igjen heilt til me oppnår luminansverdien som er så nær `0.7` som mogleg.
I iterasjon 3 av blåfargen (fargen rett til ventre for resultatet) ser me at luminansverdien er `0.34`. 

> Dette er fortsatt ikkje i nærheten av `0.7`, så me må derfor fortsetje med å "tinte" fargane.

Reviewer: ```suggestion
Dette er fortsatt ikkje i nærleiken av `0.7`, så me må derfor fortsetje med å "tinte" fargane.
```

Etter 15 iterasjonar for blåfargen, og 11 iterasjonar for grønfargen, har me funne den nærmaste luminansverdien til `0.7` som me kan med denne teknikken.

> Målet var å finne luminansverdien `0.7` for fargane, men som bildet over viser så blei resultatet `0.72` og `0.71`.

Reviewer: ```suggestion
Målet var å finne luminansverdien `0.7` for fargane, men som biletet over viser så blei resultatet `0.72` og `0.71`.
```

Dette viser greitt korleis interpoleringen fungerer, men det er ikkje nøyaktig nok for vårt formål.

> I koden vert det nytta ein meir effektiv metode, kalle binært søk, for å navigere seg gjennom fargerommet. 

Reviewer: ```suggestion
I koden vert det nytta ein meir effektiv metode, kalla binært søk, for å navigere seg gjennom fargerommet. 
```

Dette sikrar at resultatet blir meir nøyaktig og at det blir færrast mogleg iterasjonar.

## Slik blei fargeskalaen sjåande ut til slutt
Fargane blir genererte ved å finne faste luminansverdiar i RGB-fargerommet. 
Med denne metoden synest me fargane i kvar skala fungerte godt saman. 
Fargegenerering for ein fargeskala kan delast inn i to grupper: statiske grupper og ei eiga 'Base'-gruppe.

### Dei statiske gruppene
Desse gruppene bruker faste luminansverdiar i RGB-fargerommet for å definere kvart steg i skalaen.
Her vil for eksempel alle `Text Subtle` fargane alltid ha same kontrast mot `Surface Default`.

> Dette var veldig nyttig fordi då kunne me enkelt lage reglar på korleis fargane skulle brukast i systemet.

Reviewer: ```suggestion
Dette var veldig nyttig, fordi då kunne me enkelt lage reglar på korleis fargane skulle brukast i systemet.
```

 <Image
  src='/img/bloggen/fargeskalaer-rgb.png'
  alt=' '
  caption='Viser korleis dei statiske gruppene for ein fargeskala ser ut.'
  boxShadow={false}
  dataUnstyled
/>

> I bildet over ser du 4 fargeskalaer som blir generert basert på dei 4 fargane som du finn i toppen av bildet.

Reviewer: ```suggestion
I biletet over ser du 4 fargeskalaer som blir generert basert på dei 4 fargane som du finn i toppen av biletet.
```

> Me 

Reviewer: ```suggestion
```

### "Base"-gruppa
Denne gruppa fungerar litt annleis enn dei statiske gruppene. 
Fargen som blir brukt for å generere ein skala får navnet `Base Default`. Den beholder alltid same hex-kode. Dette sikrar at ein kan bevare identiteten til eksisterande visuelle profilar med definerte fargar. 

Fargane `Base Hover` og `Base Active` blir mørkare eller lysare avhengig av den relative luminansen til `Base Default` fargen.
Enkelt forklart handlar det om at desse fargane blir lysare om `Base Default` fargen er mørk, og mørkare om den er lys. 
Det er litt meir logikk som skjer her bak panseret, men det er ikkje avgjerande for å forstå korleis fargane blir generert på.
  
 <Image
  src='/img/bloggen/base-farger.png'
  alt=' '
  caption='Viser eksempel på korleis Base gruppa ser ut for 8 forskjellige fargeskalaer.'
  boxShadow={false}
  dataUnstyled
/>

> I bildet over ser du ein sekjon til venstre der `Base Hover` og `Base Active` fargane blir lysare, og ein seksjon til høgre der fargane blir mørkare.

Reviewer: ```suggestion
I biletet over ser du ein sekjon til venstre der `Base Hover` og `Base Active` fargane blir lysare, og ein seksjon til høgre der fargane blir mørkare.
```

> Måten me lagar desse fargane på er at me legger til eller trekker frå lightness frå fargen i HSLuv fargerommet.

Reviewer: ```suggestion
Måten me lagar desse fargane på er at me legger til, eller trekker frå, lightness frå fargen i HSLuv fargerommet.
```

> Grunnen til at me brukar HSLuv her er fordi me vil at fargane skal auke lysstyrke med ein fast verdi som oppleves som naturleg for menneskeauget.

Reviewer: ```suggestion
Grunnen til at me brukar HSLuv her er fordi me vil at fargane skal auke lysstyrke med ein fast verdi som opplevast som naturleg for menneskeauget.
```

Me opplevde at luminans fungerte dårleg som utgangspunkt for å seie at ein farge skulle bli 8% lysare eller mørkare. 
Etter at me fann lightness-verdien og luminans-verdien til fargen, interpolerer me fargen på same måte som forklart tidlegare.

## Vidare arbeid med fargegenereringa og oppsummering
No har me vore innom mykje spennande innan fargar og fargegenerering.
Godt jobba om du har klart å henge med så langt! 
Dette er eit komplekst tema som ofte krev ein del innsikt for å forstå fullt ut.
Me håpar det har vore nyttig for å få innsikt i korleis fargane våre blir genererte.

Vidare vil me jobbe med å få fargane endå meir balanserte på tvers av fargeskalaene.
Dei er litt for saturerte i nokre av skalaene per i dag. 

> Fordi me genererer fargar basert på kontrast, er det ikkje alle fargetonar som oppfører seg likt for det menneskelege auget.

Reviewer: ```suggestion
Sidan me genererer fargar basert på kontrast, er det ikkje alle fargetonar som oppfører seg likt for det menneskelege auget.
```

Fargen grøn er veldig dominerande i lys modus, medan blåfargen er svært framståande i mørk modus.
Me vil derfor jobbe med å jevne ut satureringa for alle fargetonane slik at dei oppleves som meir balanserte.

> Me vil og fortette å jobbe med å få mørk modus til å fungere bedre i systemet.

Reviewer: ```suggestion
Me vil og fortette å jobbe med å få mørk modus til å fungere betre i systemet.
```

> Det er spesielt dette med at enkelte fargar blir litt for saturerte me ønskjer å adressere.

Reviewer: Kan denne skrivast om?

I dag genererer me fargane i RGB-fargerommet, men me vil fortette å teste andre fargerom som OKLCH, HSLuv og LAB for å finne dei beste fargene for systemet.

Om du har spørsmål eller ønskjer å vite meir om korleis fargane våre blir genererte, 

> så kan du ta kontakt med oss på kommunikasjonskanalene våre som du finn i footeren på nettsida.

Reviewer: ```suggestion
så kan du ta kontakt med oss på kommunikasjonskanalane våre som du finn i footeren på nettsida.
```
